feat: Add JSON configuration import for Infrastructure as Code support

tvarohohlavy opened the pull request:
## feat: Add JSON configuration import for Infrastructure as Code support

### Summary
Adds the ability to configure Zerobyte entirely via a JSON configuration file, enabling Infrastructure as Code (IaC) workflows, Docker Compose deployments, and migration between instances.

### Features

#### Configuration File Support
- Load configuration from a JSON file or stdin
- Environment variable interpolation with `${VAR_NAME}` syntax for secure secret injection
- ArkType schema validation with detailed error messages and paths
- Create-only semantics: existing resources are not overwritten

#### CLI Command
```bash
bun run cli import-config --config /path/to/config.json
bun run cli import-config --stdin
bun run cli import-config --config /path/to/config.json --dry-run
```

**Options:**
| Option | Description |
|--------|-------------|
| `--config <path>` | Path to the configuration file |
| `--stdin` | Read configuration from stdin |
| `--dry-run` | Validate config without importing |
| `--json` | Output results in JSON format |
| `--log-level <level>` | Set log level (debug, info, warn, error) |
| `--overwrite-recovery-key` | Overwrite existing recovery key (only if database is empty) |

#### Supported Resources
- **Volumes**: Local directories, NFS, SMB, WebDAV, SFTP backends
- **Repositories**: Local (new or existing), S3, R2, GCS, Azure, SFTP, WebDAV, rclone backends
- **Notification Destinations**: Email, Slack, Discord, Gotify, ntfy, Pushover, Telegram, Custom (shoutrrr)
- **Backup Schedules**: Cron expressions, retention policies, include/exclude patterns, mirrors, notifications
- **Admin User**: Setup with username/password or pre-hashed password for migration
- **Recovery Key**: Setup with provided 64-character hex string

#### Import Order & Dependencies
1. Recovery key written first (required for repository encryption)
2. Volumes created
3. Repositories created (with smart duplicate detection by URL/path)
4. Notification destinations created
5. Backup schedules created (with automatic volume mounting, notification and mirror attachments)
6. Admin user created

#### Repository Import Safeguards
- Skips if repository with same URL/path already exists in DB
- For local repos: auto-detects existing restic repositories on disk
- Set `isExistingRepository: true` to import pre-existing restic repositories
- Detailed warning messages for skipped resources

#### Migration Support
- Import pre-hashed passwords (`passwordHash` field) for seamless admin migration
- Compatible with export files from PR #92

### Changes

#### New Files
- config-import.ts - ArkType validation schemas for config import
- config-import.ts - Core import logic with modular functions
- import-config.ts - CLI command implementation
- config-file-import - Complete example with docker-compose, config template, and README

#### Modified Files
- backups.dto.ts - Export `retentionPolicySchema` for reuse
- index.ts - Register `import-config` CLI command
- README.md - Added config file import to examples list

### Example Configuration

```json
{
  "recoveryKey": "${RECOVERY_KEY}",
  "volumes": [
    { "name": "my-data", "config": { "backend": "directory", "path": "/data" } }
  ],
  "repositories": [
    { "name": "s3-backup", "config": { "backend": "s3", "bucket": "backups", "accessKeyId": "${AWS_KEY}", "secretAccessKey": "${AWS_SECRET}" } }
  ],
  "backupSchedules": [
    {
      "volume": "my-data",
      "repository": "s3-backup",
      "cronExpression": "0 2 * * *",
      "retentionPolicy": { "keepLast": 7, "keepDaily": 7 },
      "notifications": ["slack-alerts"],
      "mirrors": [{ "repository": "local-mirror" }]
    }
  ],
  "notificationDestinations": [
    { "name": "slack-alerts", "config": { "type": "slack", "webhookUrl": "${SLACK_WEBHOOK}" } }
  ],
  "users": [
    { "username": "admin", "password": "${ADMIN_PASSWORD}" }
  ]
}
```

### CLI Usage

```bash
# Import from file
docker compose exec zerobyte bun run cli import-config --config /app/zerobyte.config.json

# Import from stdin
cat config.json | docker compose exec -T zerobyte bun run cli import-config --stdin

# Dry run (validate only)
docker compose exec zerobyte bun run cli import-config --config /app/zerobyte.config.json --dry-run

# JSON output for scripting
docker compose exec zerobyte bun run cli import-config --config /app/zerobyte.config.json --json
```

### Future Possible Enhancements
- Lock configuration elements against UI changes
- Visibly tag imported elements in UI
- Auto-sync between JSON config and UI changes

### Related Issues
#13

### Related PRs
Import compatible with export files from #92



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Import full configuration from JSON via a new CLI command (file or stdin) with dry-run and JSON/human-readable output.
  * Option to supply custom short IDs for volumes, repositories, and schedules (validated and checked for collisions).

* **Documentation**
  * Added detailed docs, example config, env template, docker-compose sample, and usage examples.

* **Validation & Errors**
  * Improved validation, clearer error messages, and pre-checks for existing repositories and resources.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

tvarohohlavy committed:
config import via json

tvarohohlavy committed:
avoid conflict with export PR

tvarohohlavy committed:
admin passwordHash import support

nicotsx commented:
Hello, I appreciate your contribution however I cannot afford to spend the time reviewing it as it has been AI generated. Were you able to test the implementation?

tvarohohlavy committed:
examples corrected

tvarohohlavy committed:
enhance config interpolation and validation for admin user setup

tvarohohlavy committed:
add missing env vars to docker-compose.yml

tvarohohlavy committed:
better handle importing of existing local repository config + documentation

tvarohohlavy commented on zerobyte.config.json:
```diff
@@ -0,0 +1,231 @@
+      "config": {
+        "shoutrrrUrl": "${SHOUTRRR_URL}"
+      }
+    }
```
[examples corrected](https://github.com/nicotsx/zerobyte/pull/93/commits/34123becd040bfbdcb2ae9eccfa5631803086e48)

tvarohohlavy commented on app/server/modules/repositories/repositories.service.ts:
```diff
@@ -62,7 +72,8 @@ const createRepository = async (name: string, config: RepositoryConfig, compress
 	const id = crypto.randomUUID();
-	const shortId = generateShortId();
+
+	const shortId = (config.backend === "local" && config.name) ? config.name : generateShortId();
```
[better handle importing of existing local repository config + documen…](https://github.com/nicotsx/zerobyte/pull/93/commits/1de82fa8633b8d53a85a94e445a7aefa8fd0de12)

tvarohohlavy commented on README.md:
````diff
@@ -37,6 +37,365 @@ Zerobyte is a backup automation tool that helps you save your data across multip
+        "shoutrrrUrl": "${SHOUTRRR_URL}"
+      }
+    }
+    ```
````
[examples corrected](https://github.com/nicotsx/zerobyte/pull/93/commits/34123becd040bfbdcb2ae9eccfa5631803086e48)

tvarohohlavy commented on zerobyte.config.json:
```diff
@@ -0,0 +1,231 @@
+        "privateKey": "${SFTP_PRIVATE_KEY}",
+        "path": "/backups"
+      }
+    }
```
[examples corrected](https://github.com/nicotsx/zerobyte/pull/93/commits/34123becd040bfbdcb2ae9eccfa5631803086e48)

tvarohohlavy commented on app/server/modules/lifecycle/startup.ts:
```diff
@@ -12,13 +12,204 @@ import { BackupExecutionJob } from "../../jobs/backup-execution";
+		const fs = await import("node:fs/promises");
+		const { RESTIC_PASS_FILE } = await import("../../core/constants.js");
+		if (configFileAdmin && configFileAdmin.recoveryKey) {
+			await fs.writeFile(RESTIC_PASS_FILE, configFileAdmin.recoveryKey, { mode: 0o600 });
```
[enhance config interpolation and validation for admin user setup](https://github.com/nicotsx/zerobyte/pull/93/commits/a4d1cc079111b973d67c3e63775f6afedecfc603)

tvarohohlavy commented on app/server/modules/lifecycle/startup.ts:
```diff
@@ -12,13 +12,204 @@ import { BackupExecutionJob } from "../../jobs/backup-execution";
+						const { user } = await authService.register(configFileAdmin.username, configFileAdmin.password);
+						userId = user.id;
+						logger.info(`Admin user '${configFileAdmin.username}' created from config.`);
+					}
```
[enhance config interpolation and validation for admin user setup](https://github.com/nicotsx/zerobyte/pull/93/commits/a4d1cc079111b973d67c3e63775f6afedecfc603)

tvarohohlavy commented on app/server/modules/lifecycle/startup.ts:
```diff
@@ -12,13 +12,204 @@ import { BackupExecutionJob } from "../../jobs/backup-execution";
+				}
+			} else {
+				logger.warn("Admin config missing required fields (username, password or passwordHash). Skipping automated admin setup.");
+			}
```
[enhance config interpolation and validation for admin user setup](https://github.com/nicotsx/zerobyte/pull/93/commits/a4d1cc079111b973d67c3e63775f6afedecfc603)

tvarohohlavy commented on app/server/modules/lifecycle/startup.ts:
```diff
@@ -12,13 +12,204 @@ import { BackupExecutionJob } from "../../jobs/backup-execution";
+		let configFileRepositories = [];
+		let configFileBackupSchedules = [];
+		let configFileNotificationDestinations = [];
+	    let configFileAdmin = null;
```
[enhance config interpolation and validation for admin user setup](https://github.com/nicotsx/zerobyte/pull/93/commits/a4d1cc079111b973d67c3e63775f6afedecfc603)

tvarohohlavy committed:
removed duplicate isencrypted check as its handled inside encrypt already

tvarohohlavy commented:
> **Actionable comments posted: 0**
> 
> Caution
> 
> Some comments are outside the diff and can’t be posted inline due to platform limitations.
> 
> ⚠️ Outside diff range comments (1)
> > app/server/modules/repositories/repositories.service.ts (1)> `25-48`: **Missing `isEncrypted()` guards will cause double-encryption of imported secrets.**
> > > The PR objectives state support for "detection of encrypted values via cryptoUtils.isEncrypted()" and importing both encrypted and cleartext secrets. However, `encryptConfig` unconditionally encrypts all sensitive fields without checking if they're already encrypted. When importing pre-encrypted values, this will corrupt the data by double-encrypting.
> > > Apply guards to all sensitive field encryption:
> > > ```diff
> > >  	switch (config.backend) {
> > >  		case "s3":
> > >  		case "r2":
> > > -				encryptedConfig.accessKeyId = await cryptoUtils.encrypt(config.accessKeyId);
> > > -				encryptedConfig.secretAccessKey = await cryptoUtils.encrypt(config.secretAccessKey);
> > > +			if (!cryptoUtils.isEncrypted(config.accessKeyId)) {
> > > +				encryptedConfig.accessKeyId = await cryptoUtils.encrypt(config.accessKeyId);
> > > +			}
> > > +			if (!cryptoUtils.isEncrypted(config.secretAccessKey)) {
> > > +				encryptedConfig.secretAccessKey = await cryptoUtils.encrypt(config.secretAccessKey);
> > > +			}
> > >  			break;
> > >  		case "gcs":
> > > -				encryptedConfig.credentialsJson = await cryptoUtils.encrypt(config.credentialsJson);
> > > +			if (!cryptoUtils.isEncrypted(config.credentialsJson)) {
> > > +				encryptedConfig.credentialsJson = await cryptoUtils.encrypt(config.credentialsJson);
> > > +			}
> > >  			break;
> > >  		case "azure":
> > > -				encryptedConfig.accountKey = await cryptoUtils.encrypt(config.accountKey);
> > > +			if (!cryptoUtils.isEncrypted(config.accountKey)) {
> > > +				encryptedConfig.accountKey = await cryptoUtils.encrypt(config.accountKey);
> > > +			}
> > >  			break;
> > >  		case "rest":
> > >  			if (config.username) {
> > > -				encryptedConfig.username = await cryptoUtils.encrypt(config.username);
> > > +				if (!cryptoUtils.isEncrypted(config.username)) {
> > > +					encryptedConfig.username = await cryptoUtils.encrypt(config.username);
> > > +				}
> > >  			}
> > >  			if (config.password) {
> > > -				encryptedConfig.password = await cryptoUtils.encrypt(config.password);
> > > +				if (!cryptoUtils.isEncrypted(config.password)) {
> > > +					encryptedConfig.password = await cryptoUtils.encrypt(config.password);
> > > +				}
> > >  			}
> > >  			break;
> > >  		case "sftp":
> > > -				encryptedConfig.privateKey = await cryptoUtils.encrypt(config.privateKey);
> > > +			if (!cryptoUtils.isEncrypted(config.privateKey)) {
> > > +				encryptedConfig.privateKey = await cryptoUtils.encrypt(config.privateKey);
> > > +			}
> > >  			break;
> > >  	}
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Also apply the same pattern to `customPassword` (lines 21-23):
> > > ```diff
> > >  	if (config.customPassword) {
> > > -		encryptedConfig.customPassword = await cryptoUtils.encrypt(config.customPassword);
> > > +		if (!cryptoUtils.isEncrypted(config.customPassword)) {
> > > +			encryptedConfig.customPassword = await cryptoUtils.encrypt(config.customPassword);
> > > +		}
> > >  	}
> > > ```
> 
> 🧹 Nitpick comments (1)
> > app/server/modules/repositories/repositories.service.ts (1)> `117-127`: **String-based error detection is fragile but acceptable.**
> > > The error message matching (`"config file already exists"`) could break if restic changes its error messages in future versions. The verification fallback via `snapshots()` is a reasonable approach to determine if the repository is actually usable.
> > > Consider extracting the error string to a constant for easier maintenance, or adding a comment noting the restic version this was tested against.
> > > ```diff
> > > +// Error message from restic when repository already initialized
> > > +const RESTIC_CONFIG_EXISTS_ERROR = "config file already exists";
> > > +
> > >  		if (errorStr.includes("config file already exists")) {
> > > ```
> 
> 📜 Review details

Re-encryption is not possible due to this in existing code:

https://github.com/nicotsx/zerobyte/blob/b8e30e298caf43e2c9779b369cfb66dfc397c042/app/server/utils/crypto.ts#L16

tvarohohlavy committed:
enhance repository creation logic to check for existing repositories and handle errors appropriately

tvarohohlavy committed:
Merge branch 'main' into config-import-json

tvarohohlavy committed:
Merge remote-tracking branch 'origin/main' into config-import-json

tvarohohlavy committed:
refactor the import logic
- Updated README to reflect changes in config file structure and usage.
- Modified error messages for user registration to be more generic.
- Cleaned up startup logic to utilize the new config import functionality when requested.
- Adjusted JSON config structure to support user definitions.

tvarohohlavy commented on app/server/modules/lifecycle/config-import.ts:
```diff
@@ -0,0 +1,404 @@
+			logger.warn(`Repository not created: ${err.message}`);
+		}
+	}
+}
```
it is actually using sealSecret which avoids double encryption

tvarohohlavy committed:
Merge origin/main into config-import-json

tvarohohlavy committed:
Merge remote-tracking branch 'origin/main' into config-import-json

tvarohohlavy committed:
moved documentation into separate example sub directory

tvarohohlavy committed:
Merge remote-tracking branch 'origin/main' into config-import-json

tvarohohlavy committed:
Merge remote-tracking branch 'origin/main' into config-import-json

tvarohohlavy committed:
feat(config-import): support mirrors, oneFileSystem, and optional flags

- Add mirrors support for backup schedules (copy to secondary repos)
- Add oneFileSystem flag to prevent crossing filesystem boundaries
- Support autoRemount=false for volumes (defaults to true)
- Support enabled=false for notification destinations (defaults to true)
- Optimize: move early return before dynamic imports in importBackupSchedules
- Update example config and README with new fields documentation

tvarohohlavy committed:
feat(cli): add import-config command for manual config import

Add CLI command to import configuration from file or stdin, providing
an alternative to the env-var-based automatic import on startup.

Features:
- `import-config --config <path>` to import from a mounted file
- `import-config --stdin` to import from piped input (no file mount needed)
- `import-config --dry-run` to validate config without importing

Changes:
- Add app/server/cli/commands/import-config.ts with new command
- Register importConfigCommand in CLI index
- Refactor config-import.ts: extract runImport() and add applyConfigImport()
  for direct config object import (used by CLI)
- Update docs with both import methods (env var and CLI examples)

tvarohohlavy committed:
Merge remote-tracking branch 'origin/main' into config-import-json

tvarohohlavy committed:
linting and SFTP volume support

tvarohohlavy committed:
feat: improve config import with result tracking and idempotent repository detection

- Add ImportResult type to track succeeded/warnings/errors counts across all import functions
- Implement multi-layer repository duplicate detection:
  - URL-based check (same path/bucket/endpoint already registered)
  - Restic repo check (path is already a restic repository)
  - Name-based fallback check
- Standardize logging between CLI and env var import methods with logImportSummary()
- CLI exits with code 1 when import has errors
- Remove shortId override for local repo migrations (use full path with isExistingRepository instead)
- Update example JSON and README:
  - Document that local repo path is optional (defaults to /var/lib/zerobyte/repositories)
  - Add existing-local-repo example with isExistingRepository: true
  - Add S3 endpoint field to example
  - Expand config behavior docs to explain all repository skip conditions
- Improve .gitignore to exclude all JSON except example template

tvarohohlavy committed:
Increment warning count on volume mount failure

tvarohohlavy committed:
Enhance error logging for repo existence check

Improve error handling in repository existence check.

tvarohohlavy committed:
Fix error handling in repository existence check

tvarohohlavy committed:
Merge branch 'main' into config-import-json

tvarohohlavy committed:
Merge branch 'main' into config-import-json

tvarohohlavy committed:
docs: add host-side environment variable interpolation examples for CI/CD

tvarohohlavy committed:
refactor: improve config import logging, add CLI JSON output and idempotent re-runs

- Add `--json` flag for machine-readable JSON output
- Add `--log-level` flag to control logging verbosity
- Add `skipped` counter to ImportResult for idempotent operations
- Change "already exists" conditions from warnings to skipped (info logs)
- Recovery key mismatch now errors and stops import early
- Pre-check volumes and notification destinations before creation
- Attachment functions merge missing items instead of overwriting
- Add toError() and mergeResults() helpers to reduce code duplication
- Extract readConfigJson() and createOutput() for cleaner CLI code
- Move fs/path imports to top level in config-import.ts

tvarohohlavy committed:
make the isexisting repo in location check for all repos

tvarohohlavy committed:
fix: normalize notification destination names before checking for existence

tvarohohlavy committed:
fix: normalize names for existence checks

tvarohohlavy committed:
feat: add support for provided shortId in volume, repository and backup and pass shortIds to service creation from config import

tvarohohlavy committed:
feat: add option to overwrite recovery key during config import and prevent overwrite if database is not empty

tvarohohlavy committed:
fix: replace Error with BadRequestError for invalid shortId format in repository and volume creation

tvarohohlavy committed:
refactor: remove config import from startup and update README for CLI usage

tvarohohlavy committed:
fix: remove startup import type from README

tvarohohlavy committed:
refactor: move toError function to utils/errors for reuse across modules

tvarohohlavy commented on app/server/cli/commands/import-config.ts:
```diff
@@ -0,0 +1,149 @@
+import { Command } from "commander";
+import path from "node:path";
+import fs from "node:fs/promises";
+import { toError } from "../../utils/errors";
```
Current codebase does not comply with this rule so until it does it should be probably ignored.

tvarohohlavy committed:
refactor: add ArkType schema validation for config import

- Add app/schemas/config-import.ts with ArkType schemas for upfront validation
- Replace runtime type checks with schema-based validation in config-import.ts
- Return validation errors early with detailed path and message info
- Export retentionPolicySchema from backups.dto.ts for reuse
- Update CLI to handle validation errors in both dry-run and import modes
- Fix README mirrors examples to use existing repository names

tvarohohlavy committed:
Merge branch 'main' into config-import-json

tvarohohlavy committed:
Merge branch 'main' into config-import-json

tvarohohlavy committed:
Merge branch 'main' into config-import-json

nicotsx commented:
I'm currently working on an way more opinionated version that will require less code to maintain
